### PR TITLE
26 only record in pr

### LINKED ISSUE #26: Implement user activity only in PR
Implement that the extension only
logs when the user is in a PR

### PR DESCRIPTION
**Pull request by mpsijm**
Will close #26 

Major changes:
- All classes are now both in the background page and in the content script (see `manifest.json`).
- The content script only hooks up to the DOM when it recognizes the URL to be a pull-request.
- When the DOM gets a click event, it sends a message with the data to be logged to the background page
- When the background page gets this message, it logs it to the database (no HTTP/S issues :D)
- A post to the database now only takes an `EventObject` and two callbacks (used to have a long signature)
- Renamed a bunch of things
  - `DatabaseAdaptable.log` -> `DatabaseAdaptable.post`
  - `DatabaseAdapter` -> `RESTApiDatabaseAdapter`
  - `DatabaseConsoleLogOnly` -> `ConsoleLogDatabaseAdapter`

### CODE REVIEW

File: src/octopeer-github/main/Controller.ts [hunk 1]
```diff
@@ -25,25 +26,75 @@ class Controller {
...
     /** Starts the Controller. After calling this, all event handlers are hooked to the DOM-tree. */
     public start() {
-        this.database = new DatabaseConsoleLogOnly(); // ("https://localhost:8000", 1, 1);
-        this.hookToDOM();
+        this.database = new ConsoleLogDatabaseAdapter(); // ("https://localhost:8000", 1, 1);
+        this.connectToContentScript();
         return this;
     }
 
     /** Hook the product of ElementBindings and ElementSelectors to the DOM-tree. */
```
Comment (MathiasMeuleman) [inline src/octopeer-github/main/Controller.ts]: Please use multiline JavaDoc

File: src/octopeer-github/main/DatabaseAdaptable/DatabaseAdaptable.ts [hunk 2]
```diff
@@ -11,16 +11,32 @@ type LineNumber = number;
...
+        duration: duration,
+        elementID: elementID,
+        eventID: eventID,
+        filename: filename,
+        lineNumber: lineNumber,
+        start: start,
+    };
 }
+// suppress TSLint unused-variable, because it is used elsewhere
+EventObject(new ElementID(0), new EventID(0), new Date(), 0);
```
Comment (MathiasMeuleman) [inline src/octopeer-github/main/DatabaseAdaptable/DatabaseAdaptable.ts]: Missing JavaDoc

File: src/octopeer-github/main/main.ts [hunk 3]
```diff
@@ -1,3 +1,23 @@
...
+                        success();
+                    },
+                });
+                sendResponse(`hooked to DOM (${location.href})`);
+            } else {
+                sendResponse(`did nothing (${location.href})`);
+            }
+        });
+    }
+}
```
Comment (MathiasMeuleman) [inline src/octopeer-github/main/main.ts]: Missing JavaDoc

**Comment by MathiasMeuleman** [comment]
It would be better if the Controller is split in a back-end and front-end part. This way the mutual exclusive parts are separated nicely. The main.ts will also be smaller in that case, which is also better :) (I liked the 4-line main.ts :innocent: )

**Comment by mpsijm** [comment]
Wow, I haven't even thought of that XD Will fix that later today :)

**Comment by mpsijm** [comment]
Sorry for the very elaborate commit message, I had to quick-push cause my battery was almost down :) I think that commit boils down to:
"Put the messages in JSON format and fix missing jQuery"

**Comment by MathiasMeuleman** [comment]
The splitting of the backgroud and foreground contents looks good! However I still sometimes get twice the message it is hooked to DOM, so still some double logging.

**Comment by mpsijm** [comment]
I've been able to reproduce this, thanks for noticing! I will quickly try to fix that, but as we're nearing our deadline, if this takes too much time, I'll leave this not-often-happening glitch till next sprint :)

File: src/octopeer-github/main/MainController.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            }, function() {
+                console.log("[WARN] Log to database of following object failed: ", message);
+                console.log("[WARN] Original sender: ", sender);
+            });
+            sendResponse({});
+        });
+    }
+
+    /**
+     * Only sends a message to a tab if they contain the correct URL.
```
Comment (mdingjan) [inline src/octopeer-github/main/MainController.ts]: if it contains the correct URL?

File: src/octopeer-github/content/ContentController.ts [hunk 5]
```diff
@@ -88,22 +87,54 @@ class Controller {
...
+    private connectToBackgroundPage() {
+        const self = this;
+        if (chrome.runtime.onMessage.hasListeners()) {
+            return;
+        }
+
+        chrome.runtime.onMessage.addListener(function (request, sender, sendResponse) {
+            if (request.hookToDom) {
+                self.hookToDOM({
+                    post: function (data: EventObject, success: Callback, failure: Callback) {
```
Comment (mdingjan) [inline src/octopeer-github/content/ContentController.ts]: What do you use failure for? Just wondering :)

Comment (mpsijm) [inline src/octopeer-github/content/ContentController.ts]: `failure` is part of the Interface of `DatabaseAdaptable` and should be called whenever the post to the database failed :) I'd better document that :P

File: src/octopeer-github/main/MainController.ts (hunk 4 above)
Comment (mpsijm) [inline src/octopeer-github/main/MainController.ts]: Yes, which is checked by the regular expression a few lines ahead :) (more documentation needed, will add this XD)

Comment (mdingjan) [inline src/octopeer-github/main/MainController.ts]: I was actually wondering why you used 'they', as you're checking the URL of one tab right?

Comment (mpsijm) [inline src/octopeer-github/main/MainController.ts]: oh yeah, right. XD Will change that as well :)

**Comment by mdingjan** [comment]
Thanks for resolving my feedback :)
You did a great job on the implementation of this issue. :)

The SAT are happy and I'm happy :)

Will merge!

**Pull request merged**